hoomd/hpmc/ShapeMoves.h

```diff
@@ -0,0 +1,870 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#ifndef _SHAPE_MOVES_H
+#define _SHAPE_MOVES_H
+
+#include "GSDHPMCSchema.h"
+#include "Moves.h"
+#include "ShapeUtils.h"
+#include "hoomd/extern/quickhull/QuickHull.hpp"
+#include <Eigen/Dense>
+#include <hoomd/Variant.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace hoomd
+    {
+
+namespace hpmc
+    {
+
+template<typename Shape> class ShapeMoveBase
+    {
+    public:
+    ShapeMoveBase(std::shared_ptr<SystemDefinition> sysdef,
+                  std::shared_ptr<IntegratorHPMCMono<Shape>> mc)
+        : m_mc(mc), m_det_inertia_tensor(0), m_sysdef(sysdef)
+        {
+        m_ntypes = this->m_sysdef->getParticleData()->getNTypes();
+        m_volume.resize(m_ntypes, 0);
+        }
+
+    virtual ~ShapeMoveBase() {};
+
+    //! prepare is called at the beginning of every update()
+    virtual void prepare(uint64_t timestep)
+        {
+        throw std::runtime_error("Shape move function not implemented.");
+        }
+
+    //! construct is called for each particle type that will be changed in update()
+    virtual void update_shape(uint64_t,
+                              Scalar& stepsize,
+                              const unsigned int&,
+                              typename Shape::param_type&,
+                              hoomd::RandomGenerator&)
+        {
+        throw std::runtime_error("Shape move function not implemented.");
+        }
+
+    //! retreat whenever the proposed move is rejected.
+    virtual void retreat(uint64_t timestep)
+        {
+        throw std::runtime_error("Shape move function not implemented.");
+        }
```

Comment:
Make pure virtual functions if you require inheriting classes to implement their own logic.